internal/rpcserver/rpcserver.go

```diff
@@ -1517,7 +1521,31 @@ func handleExistsAddress(_ context.Context, s *Server, cmd interface{}) (interfa
 			err)
 	}
 
-	exists, err := s.cfg.ExistsAddresser.ExistsAddress(addr)
+	// Ensure the exists address index is synced.
+	existsAddrIndex := s.cfg.ExistsAddresser
+	tHeight, tHash, err := existsAddrIndex.Tip()
+	if err != nil {
+		return nil, rpcInternalError(err.Error(), "Tip")
+	}
+
+	chain := s.cfg.Chain
+	if chain.BestSnapshot().Height > (tHeight + 5) {
+		msg := fmt.Sprintf("%s: index not synced", existsAddrIndex.Name())
+		return nil, rpcInternalError(msg, "Sync")
+	}
+
+sync:
+	for !chain.BestSnapshot().Hash.IsEqual(tHash) {
+		select {
+		case <-time.After(syncWait):
+			msg := fmt.Sprintf("%s: index not synced", existsAddrIndex.Name())
+			return nil, rpcInternalError(msg, "Sync")
+		case <-existsAddrIndex.WaitForSync():
+			break sync
+		}
```

Comment:
Blocking in the RPC handler ensures the response i guaranteed to be accurate to the node's current chain state and isn't being derived from stale data. We are also blocking for a maximum of `syncWait` here (3 seconds max currently). If the sync takes too long an error will be returned.

Here is a snippet from `handleGetRawTransaction`:

```go
sync:
for !chain.BestSnapshot().Hash.IsEqual(tHash) {
	select {
	case <-time.After(syncWait):
		msg := fmt.Sprintf("%s: index not synced", txIndex.Name())
		return nil, rpcInternalError(msg, "Sync")
	case <-txIndex.WaitForSync():
		break sync
	}
}

...
```

Reply:
I'm a bit torn on this.  I generally agree with @rstaudt2 that blocking in RPC handlers is typically not super desirable because it results in a laggy UX and can even break some speed guarantees that things such as the [Decred Rosetta middleware, dcrros](https://github.com/decred/dcrros) have to make.

On the other hand, callers, including the aforementioned `dcrros`, currently have the expectation that they can make queries based on the current state of the chain once they receive a `blockconnected` notification.  Without some logic like this, all of those callers who have relied on that implicit assumptions for 5+ years will suddenly break.

Along those same lines, something else that needs to be checked here is that there isn't any **internal** code in dcrd that relies on such behavior.  I don't believe there is, but it's possible something in the fee estimator code might.

In any case, given the above, I think the main two choices are to either:

a) Keep this, admittedly rather hacky, logic to ensure continuity; or
b) Separate blockchain notifications from the ultimate RPC notifications where the latter are not sent until continuity is achieved.

I think that ultimately (b) might be a better option, but it will be tricky to implement properly and probably should be separate from this PR.  For example, such logic would need to deal with cases where there are no indexers active and thus can send the notification immediately and cases where maybe some of the indexers are caught up and others aren't, among others.

One benefit of the approach in this PR that is kind of nice which option (b) wouldn't provide is that it won't block RPCs that don't care about certain indexes even if those indexes are not yet current.  For example, `getrawtransaction` won't block if the `txindex` is already synced and the `addrindex` is still working.

Reply:
> On the other hand, callers, including the aforementioned dcrros, currently have the expectation that they can make queries based on the current state of the chain once they receive a blockconnected notification. Without some logic like this, all of those callers who have relied on that implicit assumptions for 5+ years will suddenly break.

The current logic will break this assumption if the delay to asynchronously update the indexes is greater than the configured wait time (currently 3 seconds).  It's probably not likely to hit this case since we are limiting the channel buffer size, but something to consider depending on how far ahead of the indexers the chain can get during IBD.

Outside of that, I agree with your assessment and that it's probably best to keep it as is.